### DISCUSSION TITLE
Updating artifacts process for CocoaPods doesn't take account for the installation with bundler

### DISCUSSION THREAD

**Discussion by ainame** [renovatebot/renovate#19991] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.110.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

N/A

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I was trying to set up a self-hosted Renovate running on GitHub Actions or [Bitrise](https://bitrise.io/) for our mobile app projects. 

While testing Renovate with iOS project using CocoaPods, I noticed that updating artefacts process wasn't able to find `pod` command. This is because our project installs `cocoapods` gem via bundler and it doesn't install gems globally since we set `BUNDLE_PATH` internally. So even though `pod` command is installed correctly before running `renovate` command, `renovate` can fail to run the command _unless you execute it with `bundle exec` prefix_.

Looking at the source code and spec, it doesn't look like `renovate` executes `pod` command with `bundle exec`.

https://github.com/renovatebot/renovate/blob/8e8cbba0e62ac5ccac0a24de295a468c5c2101b8/lib/modules/manager/cocoapods/artifacts.ts#L72

https://github.com/renovatebot/renovate/blob/8e8cbba0e62ac5ccac0a24de295a468c5c2101b8/lib/modules/manager/cocoapods/artifacts.spec.ts#L291

My suggestion is that `updateArtifacts` method should take account of `cocoapods` gem installed via Bundler. It can be detected if `renovate` checks `Gemfile.lock`. Or maybe `renovate` can run `bundle exec pod install` first assuming `cocoapods` is installed by `bunlder` and if it fails, it can fallback naive `pod install` command that assumes `cocoapods` gem is installed globally.

To replicate the issue, you can create an iOS app with CocoaPods integrations and `Gemfile` to install `cocoapods` gem. You also need to have this `.bundle/config` to make sure `cocoapods` isn't installed globally.

```
---
BUNDLE_PATH: "vendor/bundler"
```

### Relevant debug logs

<details><summary>Logs</summary>

```
 INFO: Repository started (repository=XXXXXXXXXX)
       "renovateVersion": "34.110.0"
 INFO: Dependency extraction complete (repository=XXXXXXX, baseBranch=main)
       "stats": {
         "managers": {
           "bundler": {"fileCount": 1, "depCount": 22},
           "cocoapods": {"fileCount": 1, "depCount": 17},
           "github-actions": {"fileCount": 2, "depCount": 4},
           "mint": {"fileCount": 1, "depCount": 10},
           "ruby-version": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 6, "depCount": 54}
       }
 INFO: Branch created (repository=XXXXXXXXXX, branch=renovate/mbprogresshud)
       "commitSha": "2fe8d421c05413279a1ad4dc1dcfcd3ed77ccd70"
 INFO: PR created (repository=XXXXXXXXXX, branch=renovate/mbprogresshud)
       "pr": 19010,
       "prTitle": "Update dependency MBProgressHUD to v1.2.0"
 WARN: artifactErrors (repository=XXXXXXXXXX, branch=renovate/mbprogresshud)
       "artifactErrors": [
         {
           "lockFile": "Podfile.lock",
           "stderr": "rbenv: pod: command not found\n\nThe `pod' command exists in these Ruby versions:\n  2.6.9\n  2.7.5\n  2.7.6\n  3.1.3\n\n"
         }
       ]
 INFO: Comment added (repository=XXXXXXXXXX, branch=renovate/mbprogresshud)
       "issueNo": 19010,
       "topic": "⚠ Artifact update problem"
 INFO: Repository finished (repository=XXXXXXXXXX)
       "cloned": true,
       "durationMs": 56430
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
converted to discussion as this isn't a bug.

**Comment by ainame** [comment]
I think this could have been a feature request, then. I will try to support this🤔

**Reply by viceice** [reply]
can't you simply add the bundler exec path to the global `PATH` before running renovate. then it should simply work.

**Reply by ainame** [reply]
Good point. I tried that now. It didn't work. I believe it won't work that simply. This is because of how A Ruby script/executable with a bundler loads other libraries. `pod` command couldn't even load other gems that `pod` depends on internally. We would have to tweak other environment variables to make all work fine.

`bundle exec` seems to modify the environment on demand. I think asking Renovate users to use such workarounds to replicate `bundle exec` behaviour isn't ideal.
https://bundler.io/v2.4/man/bundle-exec.1.html#ENVIRONMENT-MODIFICATIONS

**Reply by viceice** [reply]
create a simple shell script in path wich calls the real command?

**Reply by ainame** [reply]
Thanks for the suggestion! I think at least I can work around the issue for now, but I don't think that's an ideal solution for less experienced engineers who really need the help of `renovate`🤔 Would you mind if I submit a PR to support `bundle exec` prefix depending on user's environment?

**Reply by viceice** [reply]
why you need your own pod version at all? doesn't renovate detect the appropriate version and install it on the fly?

**Reply by ainame** [reply]
I rechecked the issue list now. According to the issue below, such functionality isn't implemented yet for CocoaPods. 
https://github.com/renovatebot/renovate/issues/6007

But I can see your point that we don't have to use `bundle exec` if Renovate installs cocoapods based on `Gemfile.lock` or `Podfile.lock` before updating artefacts.

**Reply by ainame** [reply]
I dug into the source code and document. Now I have a better idea of how Renovate works. I was running renovate npm CLI directly on macOS VM (Bitrise) because I wanted to resolve macOS-dependent issues while using Renovate, but that didn't get the dynamic install. I could simply work around by manually install `gem install cocoapods -v `cat Podfile.lock | tail -1 | cut -d : -f2` for now😅 

> Additionally, when Renovate is run inside a container built using [containerbase](https://github.com/containerbase), such as the official Renovate images on Docker Hub, then binarySource=install can be used. This mode means that Renovate will dynamically install the version of tools available, if supported.

https://github.com/renovatebot/docker-buildpack/blob/d14b1584d8fe446e9766ddd1aef0d3831dc5f6af/src/base/bin/install-tool
